articles/serializers.py

```diff
@@ -40,10 +48,183 @@ class Meta:
+                return {
+                    **representation,
+                    "favorited": True,
+                    "unfavorited": False,
+                }
+            if request.user in instance.dislikes.all():
+                return {
+                    **representation,
+                    "favorited": False,
+                    "unfavorited": True,
+                }
+            return {**representation, "favorited": False, "unfavorited": False}
+        return representation
+
+
+class ArticleFavoriteSerializer(serializers.ModelSerializer):
+    likes = UserSerializer(many=True, required=False, read_only=True)
+    dislikes = UserSerializer(many=True, required=False, read_only=True)
+    tags = TagListSerializerField()
+    likes_count = serializers.SerializerMethodField()
+    dislikes_count = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Article
+        fields = (
+            "lookup_id",
+            "slug",
+            "title",
+            "description",
+            "image",
+            "body",
+            "tags",
+            "is_hidden",
+            "likes_count",
+            "dislikes_count",
+            "likes",
+            "dislikes",
+            "created_at",
+            "updated_at",
+            "author",
+        )
+        read_only_fields = [
+            "created_at",
+            "updated_at",
+            "author",
+            "likes",
+            "dislikes",
+        ]
+
+    def get_likes_count(self, instance: Any) -> Any:
+        return instance.likes.count()
+
+    def get_dislikes_count(self, instance: Any) -> Any:
+        return instance.dislikes.count()
+
+    def update(self, instance: Any, validated_data: Any) -> Any:
+        """update the likes of an article"""
+        request = self.context.get("request")
+
+        if request.user in instance.likes.all():
+            instance.likes.remove(request.user)
+            return instance
+        if request.user in instance.dislikes.all():
+            instance.dislikes.remove(request.user)
+        instance.likes.add(request.user)
+        return instance
+
+    def to_representation(self, instance: Any) -> Any:
+
+        """check if the user liked or disliked an article and return favorited and unfavorited status"""
+        request = self.context.get("request")
+        representation = super().to_representation(instance)
+
+        if request.user in instance.likes.all():
+            return {
+                **representation,
+                "favorited": True,
+                "unfavorited": False,
+            }
+        if request.user in instance.dislikes.all():
+            return {
+                **representation,
+                "favorited": False,
+                "unfavorited": True,
+            }
+        return {**representation, "favorited": False, "unfavorited": False}
+
+
+class ArticleUnFavoriteSerializer(serializers.ModelSerializer):
```

Comment:
I feel like you are repeating a alot of functionality of ArticleFavoriteSerializer here, can they not be combined into one? you know DRY concept

Reply:
I tried it but the overrides and nesting complicated the logic too much. At times one thing would not just work right

Reply:
Yeah but you can then have a parent serializer that has the common functions and properties and then inherit them and change the functionaluity you need

Reply:
Done